### PR TITLE
Support .Net Core analysis in container

### PR DESCRIPTION
**Pull request by johnwatson484**
https://eaflood.atlassian.net/browse/PSD-477

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,27 @@
+ARG PARENT_VERSION=1.0.1-dotnet3.1
+
+FROM defradigital/dotnetcore-development:${PARENT_VERSION} AS development
+
+# Install Sonar Scanner, Coverlet and Java (required for Sonar Scanner)
+USER root
+RUN apk --no-cache add openjdk11 --repository=http://dl-cdn.alpinelinux.org/alpine/edge/community
+USER dotnet
+RUN dotnet tool install --global dotnet-sonarscanner
```
Comment (markysoft) [inline Dockerfile]: I don't know how important image size is, but if so the run commands could be combined to create a single layer by concatenating the commands with `&&`

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+RUN dotnet tool install --global coverlet.console
+ENV PATH="$PATH:/home/dotnet/.dotnet/tools"
+
+# Map args to env vars
+ARG PARENT_VERSION
+ENV SONAR_HOST_URL https://sonarcloud.io
+ENV SONAR_ORGANIZAION_KEY defra
```
Comment (markysoft) [inline Dockerfile:16]: Could do with being consistent on how ENV vars are set, some have an `=` before the value and others do not.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ENV SONAR_PR_PROVIDER=GitHub
+ENV FIX_COVERAGE_REPORT=true
+ENV RUN_TESTS=false
+
+LABEL uk.gov.defra.ffc.parent-image=defradigital/dotnetcore-development:${PARENT_VERSION}
+
+COPY --chown=dotnet:dotnet ./scripts .
+RUN install -d -o dotnet -g dotnet /home/dotnet/project/
+RUN install -d -o dotnet -g dotnet /home/dotnet/working/
```
Comment (markysoft) [inline Dockerfile]: could combine these two runs with `&&` and perhaps a `\` to split the lines to make it more readable if image size is a concern.

File: README.md [hunk 4]
```diff
@@ -1,5 +1,55 @@
-# .Net Core Sonar
-Docker image for analysing .Net Core services running with in containers.
+# .Net Core SonarClound Analysis
```
Comment (markysoft) [inline README.md]: Bit of a nitpick, but `.NET` tends to be written in all caps

File: README.md [hunk 5]
```diff
@@ -1,5 +1,55 @@
...
+
+## Prerequisites
+- SonarCloud project
+- SonarCloud authentication token
+- `.sln` file must exist in root workspace folder as it is a dependency of SonarScanner
+- if code coverage is required, then unit test project must include [coverlet.msbuild](https://www.nuget.org/packages/coverlet.msbuild) Nuget package
+
+## Usage
+### Main branch analysis
+To analysis the main branch of a repository run the following command in the working directory of the solution.
```
Comment (markysoft) [inline README.md]: I think `analysis` should be `analyse` here

File: README.md [hunk 6]
```diff
@@ -1,5 +1,55 @@
...
+By default, the image will assume that there is an existing `coverage.opencover.xml` file in the working directory, `./test-output`.  The image will use this file to report coverage to SonarCloud.  
+
+If it is safe for your service to run tests within the container as part of the analysis, this can be done by passing in the following additional environment variables to the `docker run` commands above.  
+
+```
+-e RUN_TESTS=true
+-e FIX_COVERAGE_REPORT=false
+```
+
+**Note** in order to succesfully keep the relationship between an existing coverage report and the SonarScanner analysis results, the analysis process will remap file locations within this report.  If there is no existing coverage report or you wish to run your own tests this should be disabled by setting the `FIX_COVERAGE_REPORT` value to `false`.
```
Comment (markysoft) [inline README.md]: Typo in `successfully`

**Review by markysoft — CHANGES_REQUESTED** [review]
Looks good, a suggestion for changes if the image needs to be a bit smaller, and a couple of typos to fix.

File: Dockerfile (hunk 1 above)
Comment (johnwatson484) [reply Dockerfile]: > Looks good, a suggestion for changes if the image needs to be a bit smaller, and a couple of typos to fix.

Thanks, the SDK image is already pretty big so anything that can be done to restrict it's size would be wise.  I've made your suggested changes.

**Comment by johnwatson484** [comment]
> Looks good, a suggestion for changes if the image needs to be a bit smaller, and a couple of typos to fix.

Thanks, the SDK image is already pretty big so anything that can be done to restrict it's size would be wise. I've made your suggested changes.

**Review by markysoft — APPROVED** [review]
LGTM

**Review by markysoft — APPROVED** [review]
SLGTM

**Review by markysoft — APPROVED** [review]
Still looks good to me!

**Pull request merged**